File: src/java/TransXML.java
import java.io.File;
import javax.xml.XMLConstants;
import javax.xml.transform.*;
import javax.xml.transform.stream.StreamSource;
import javax.xml.transform.stream.StreamResult;
import javax.xml.validation.*;

public class TransXML {

    public static void main(String[] args) {
        // File paths (adjust if needed)
        String xmlPath = "Tours.xml";
        String xsdPath = "tours.xsd";
        String xsltPath = "UpcominEvents.xsl";
        String outputPath = "output/outputJava.html";

        // Validate XML against XSD
        if (validateXML(xmlPath, xsdPath)) {
            System.out.println("✅ XML is valid against the schema.");
        } else {
            System.out.println("❌ XML validation failed. Fix errors before transforming.");
            return;
        }

        // Apply XSLT transformation
        try {
            TransformerFactory factory = TransformerFactory.newInstance();
            Source xslt = new StreamSource(new File(xsltPath));
            Transformer transformer = factory.newTransformer(xslt);

            Source xml = new StreamSource(new File(xmlPath));
            Result output = new StreamResult(new File(outputPath));

            transformer.transform(xml, output);
            System.out.println("✅ Transformation complete. Output saved to: " + outputPath);
        } catch (Exception e) {
            System.out.println("❌ Error during transformation: " + e.getMessage());
            e.printStackTrace();
        }
    }

    // validate XML against XSD
    public static boolean validateXML(String xmlPath, String xsdPath) {
        try {
            SchemaFactory schemaFactory = SchemaFactory.newInstance(XMLConstants.W3C_XML_SCHEMA_NS_URI);
            Schema schema = schemaFactory.newSchema(new File(xsdPath));
            Validator validator = schema.newValidator();
            validator.validate(new StreamSource(new File(xmlPath)));
            return true;
        } catch (Exception e) {
            System.out.println("❌ Validation error: " + e.getMessage());
            return false;
        }
    }
}

File: src/python/TransXML.py
import sys
import os
from lxml import etree

XML_PATH = "Tours.xml" 
# ----------------------------------------------------

if len(sys.argv) != 4:
    print("Usage: python TransXML.py <XSLT_FILE> <OUTPUT_FILE> <OUTPUT_FOLDER>")
    sys.exit(1)

XSLT_PATH = sys.argv[1]
OUTPUT_FILE = sys.argv[2]
OUTPUT_FOLDER = sys.argv[3]


def transform_xml(xml_path, xslt_path, output_path):
    """Applies an XSLT stylesheet to an XML file."""
    try:
        xml_doc = etree.parse(xml_path)
        xslt_doc = etree.parse(xslt_path)
        
        transform = etree.XSLT(xslt_doc)
        result_tree = transform(xml_doc)
        
        # Save output. 
        with open(output_path, 'wb') as f:
            f.write(result_tree)
            
        print(f"SUCCESS: {xslt_path} -> {output_path}")
    except Exception as e:
        print(f"FAILURE: Error during transformation {xslt_path}: {e}")


if __name__ == "__main__":
    output_path = os.path.join(OUTPUT_FOLDER, OUTPUT_FILE)
    transform_xml(XML_PATH, XSLT_PATH, output_path)